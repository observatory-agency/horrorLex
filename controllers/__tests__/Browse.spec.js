const BrowseController = require('../Browse');
const Browse = require('../../lib/Browse');

jest.mock('../../lib/Browse');
jest.mock('../../lib/Mongo', () => ({
  db: {
    collection: jest.fn(() => ({
      aggregate: jest.fn(),
      createIndex: jest.fn(),
      drop: jest.fn(),
      findOne: jest.fn(),
      find: jest.fn(),
      insertMany: jest.fn(),
    })),
  },
}));

describe('BrowseController', () => {
  let browse;
  beforeEach(() => {
    Browse.mockImplementation(() => ({
      byChar: jest.fn(() => []),
    }));
    browse = new BrowseController();
  });

  describe('constructor', () => {
    it('should set its templates', () => {
      expect(browse.template).toMatchSnapshot();
    });
  });

  describe('get', () => {
    describe('without a char param', () => {
      it('should call res.render with "get" template and results', async () => {
        const renderMock = jest.fn();
        const resMock = { render: renderMock };
        const reqMock = { params: { view: 'film' } };
        const nextMock = jest.fn();
        await browse.get(reqMock, resMock, nextMock);
        expect(renderMock).toHaveBeenCalledWith(
          browse.template.get, { results: [], view: 'film' },
        );
      });
    });
    describe('with a char param', () => {
      it('should call res.render with "get" template and results', async () => {
        const renderMock = jest.fn();
        const resMock = { render: renderMock };
        const reqMock = { params: { char: 'a', view: 'film' } };
        const nextMock = jest.fn();
        await browse.get(reqMock, resMock, nextMock);
        expect(renderMock).toHaveBeenCalledWith(
          browse.template.get, { results: [], view: 'film' },
        );
      });
    });
    describe('with an invalid view param', () => {
      it('should call res.renderStatus with 404', async () => {
        const sendStatusMock = jest.fn();
        const resMock = { sendStatus: sendStatusMock };
        const reqMock = { params: { char: 'a', view: 'articles' } };
        const nextMock = jest.fn();
        await browse.get(reqMock, resMock, nextMock);
        expect(sendStatusMock).toHaveBeenCalledWith(404);
      });
    });
    describe('other errors', () => {
      it('should call "next" on errors', async () => {
        const error = new Error();
        const resMock = jest.fn();
        const reqMock = { params: { char: 'a', view: 'film' } };
        const nextMock = jest.fn();
        Browse.mockImplementation(() => ({
          byChar: jest.fn(() => { throw error; }),
        }));
        await browse.get(reqMock, resMock, nextMock);
        expect(nextMock).toHaveBeenCalledWith(error);
      });
    });
  });
});
