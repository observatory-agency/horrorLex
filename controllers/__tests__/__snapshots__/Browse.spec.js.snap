// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrowseController constructor should set its templates 1`] = `
Object {
  "get": "browse.hbs",
}
`;
